contrib_bots/lib/spotify.py

```diff
@@ -0,0 +1,124 @@
+class SpotifyHandler(object):
+    '''
+    This bot will allow users to directly search
+    for songs, artists, and albums using the Spotify API.
+    It can be called using `@spotify song/artist/album <your_search_query>`
+    '''
+    def usage(self):
+        return '''
+        This bot will allow users to directly search
+        for songs, artists, and albums using the Spotify API.
+        It can be called using `@spotify song/artist/album <your_search_query>`
+        This will return the top 10 search results from Spotify for either
+        song/artist/album.
+            '''
+
+    def triage_message(self, message):
+        # return True iff we want to (possibly) response to this message
+        original_content = message['content']
+        # This next line of code is defensive, as we
+        # never want to get into an infinite loop of posting follow
+        # ups for own follow ups!
+        if message['display_recipient'] == 'spotify':
+            return False
+        is_spotify = original_content.startswith('@spotify')
+
+        return is_spotify
+
+    def handle_message(self, message, client, state_handler):
+
+        def build_data(query_type, original_content):
+            if query_type = "song":
+                encoded_message = urllib.parse.quote(original_content[5:])
+                url = BASE_URL+encoded_message+"&type=track"+"&limit=10"
+                return urllib.request.urlopen(url).read()
+            elif query_type = "album":
+                encoded_message = urllib.parse.quote(original_content[6:])
+                url = BASE_URL+encoded_message+"&type=album"+"&limit=10"
+            elif query_type = "artist":
+                encoded_message = urllib.parse.quote(original_content[7:])
+                url = BASE_URL+encoded_message+"&type=artist"+"&limit=10"
+            return urllib.request.urlopen(url).read()
+
+        def artist_content(artist):
+            artist_name = artist['name']
+            artist_url = artist['external_urls']['spotify']
+            return [artist_name, artist_url]
+
+        def album_content(album):
+            album_name = album['name']
+            album_url = album['external_urls']['spotify']
+            artist_name = album['artists'][0]['name']
+            artist_url = album['artists'][0]['external_urls']['spotify']
+            return [album_name, album_url, artist_name, artist_url]
+
+        def track_content(track):
+            song_name = track['name']
+            song_url = track['external_urls']['spotify']
+            album_name = track['album']['name']
+            album_url = track['album']['external_urls']['spotify']
+            artist_name = track['artists'][0]['name']
+            artist_url = track['artists'][0]['external_urls']['spotify']
+            return [song_name, song_url, album_name, album_url, artist_name, artist_url]
+
+        BASE_URL = 'https://api.spotify.com/v1/search?q='
```

Comment:
Then I'd need to call it back into the function and since it's only being used within the function I don't think it makes sense moving it above the function. Unless that's the format we're following.